### ISSUE TITLE
remap bug when original image is larger than 32767 in size

### ISSUE THREAD

**Issue by saedrna** [opencv/opencv#7544]
##### System information (version)

- OpenCV => 3.1:
- Operating System / Platform => Windows 64 Bit:
- Compiler => Visual Studio 2013:

##### Detailed description

When trying to do remap with a quite large image, whose size is larger than 32767 (max size of int16), remap function will not act normally. This is because that in implementation, remap uses int16 as the coordinate data, as ```Mat _bufxy(brows0, bcols0, CV_16SC2), _bufa;```.  So when using it, the rows larger than 32767 will be clamped to 32767 and generate the following striped results.

![image](https://cloud.githubusercontent.com/assets/3850191/19643270/13eda790-9a1c-11e6-98c9-39680c5c39b9.png)

**Comment by pssp123** [comment]
Hi! Have you found the resolution to this bug?

**Comment by saedrna** [comment]
I spent a week to locate this bug and finally figured out it is the problem of opencv. Then I turn to use ipp for the remap. Ipp accepts two float maps for coordinating, and results are normal

**Comment by kinchungwong** [comment]
All of the remap functions in OpenCV imgwarp.cpp had been converted into using 16-bit integer coordinates internally, so unfortunately there is no way back.

For certain classes of remap problems, if the coordinate transform is invertible (i.e. it is mathematically trivial to convert between the forward mapping and backward mapping, and that it is computationally trivial to populate a large matrix of floating-point mapping coordinates), then you can use the tiling approach.

Sample code for tiling approach: https://gist.github.com/kinchungwong/141bfa2d996cb5ae8f42

In the tiling approach, one divides the output image into tiles, and generate each tile using remap. For each output tile, there is a corresponding area of input (not necessary rectangular, possibly warped or even curved depending on the type of coordinate transformations), but you can always find a axis-aligned rectangular bounding box for this area of input. The tile size need to be chosen so that every input area bounding box does not exceed 32767 bytes per row (for an earlier SIMD-related bug) or 32767 pixels in any dimensions (for the current bug caused by the use of signed 16-bit integers as coordinates).

During processing, each input area need to be copied into a temporary matrix satisfying the conditions (maximum bytes per row, and maximum dimensions), before passing the input matrix to remap.

**Comment by saedrna** [comment]
@kinchungwong , the tiling strategy is good, but I don't think it should be included in opencv, because it's not quite general. In my opinion, if someone wanted to do tiling, he should write his own tiling functions (and the affine transform is also an specific function).

For opencv, I think, at least it can check the size of the input and output images and throw an exception or  give an assertion failure if the size exceeds the 32767. This will let the user know it's the opencv problem and save time in debugging (one week for me...).

Better, it may write another simple wrapper for ```ipp``` with two float maps, and turn to use ```ipp``` when this is the case.

**Comment by alalek** [comment]
> at least it can check the size of the input and output images and throw an exception or give an assertion failure if the size exceeds the 32767

:+1:

**Comment by souch55** [comment]
Hi alalek i agree with you size can be checked with assertion failure, but if some one wants use large Image and remap them then we have to do conversion shorts with int

**Comment by kinchungwong** [comment]
Future plan for this issue: 

 * [Evolution Proposal OE-17, New Filter Engine](https://github.com/opencv/opencv/wiki/OE-17.-New-Filter-Engine)
 * Discussions for OE-17: #11012 

Either this OE-17, or an amended one, will cover the ```cv::remap``` use case, which will use tiling to resolve any limitations inherent in the low-level algorithm implementation. This enhancement will be parallelizable.

**Comment by alalek** [comment]
Not sure, that remap problem correlates with FilterEngine.
Some part of this is related to `CV_16SC2` datatype for maps (we need support to something like CV_32SC2 in this case and have some checks for CV_32FC2).

**Comment by kinchungwong** [comment]
@alalek Yes, I agree that the issue with ```remap``` using ```CV_16SC2``` is also a blocker. Both issues (plus some others) needs to be fixed. Do you think this altogether (```remap``` for large images) can be converted into a Evolution Proposal, or to keep it as a normal issue? It will probably require a rewrite.

**Comment by YoannNicod** [comment]
What is the current situation concerning this issue ?
I see it is still open, what could be expected and in what time frame ?
I agree with @saedrna that an exception should be thrown if the size is too great. I encountered this problem in my software in production, and I spend some time to figure out the cause of the strange beheviour I had. Would be great to inform.

Also, warpAffine() is concerned and there is nothing in the documentation saying so.

**Comment by alalek** [comment]
> exception should be thrown

It is already done (~2 years ago, see #7601).

**Comment by YoannNicod** [comment]
Thank you, I was using a previous version of OpenCV, that's why I did not see it.
Do you know if any action on the original issue is planned ?

**Comment by jjahanip** [comment]
I am using the cv2.warpPerspective in opencv-python and I have this issue on large images (>32767 in dimension).
Anyway to fix the issue?

**Comment by YoannNicod** [comment]
> I am using the cv2.warpPerspective in opencv-python and I have this issue on large images (>32767 in dimension).
> Anyway to fix the issue?

Unfortunately no. Except if this has been fixed since my last post, there is no solution. I handled the problem with a condition : if the dimension is < 32767 px, I used OpenCV (rotation) function, else I implemented the algorithm myself, which was of course slower...
This issue is serious in my opinion.

**Comment by nh2** [comment]
Why does `remap()` not just take a lambda?

The mapping is just a pure function, there seems to be no need to represent it explicitly in memory.

Getting rid of the mapping as a `Mat` would also reduce memory use. Currently the mapping has to be explicitly represented as a `Mat` of output dimensions (and that in practice with 2*float = 8 Bytes per pixel, which is almost 3x larger than typical output 3 Bytes per pixel for RGB).

I suppose (but haven't checked it) that OpenCV uses a lookup table because some interpolation methods need to look up the same coordinates repeatedly, such as LANCZOS4 with its 8x8 neighborhood.

Still, even that could be handled with a much smaller sweeping buffer of e.g. `width*8` instead of full `width*height`. Or it could just offer to always use the pure functions and use no additional memory.
